File: frontend/src/lib/recoveryPhrase.ts
/**
 * Recovery phrase decode/validate utilities (frontend port of
 * `app/security/recovery_phrase.py`, ADR-0003 / #114).
 *
 * Pure functions, no I/O at call time. The 2048-word BIP39 English
 * wordlist is vendored at `bip39-wordlist.txt` and imported as raw
 * text via Vite's `?raw` loader so this module works offline and
 * needs no build step beyond the standard Vite pipeline.
 *
 * Terminology discipline (ADR-0003): the user-facing concept is
 * "recovery phrase". This file uses BIP39 internally only because
 * that is the upstream wordlist's actual name; nothing exported
 * here surfaces "BIP39", "mnemonic", or "seed" to the operator.
 *
 * The frontend uses these helpers to:
 *   - validate a 24-word phrase BEFORE submitting to /auth/recover
 *     so a typo surfaces as a precise "word N is not recognised"
 *     message rather than the backend's generic 400
 *   - confirm the checksum byte locally so a phrase with the right
 *     shape but a transcription error fails fast
 *
 * What this file deliberately does NOT do:
 *   - encode a phrase from entropy (the frontend never holds the
 *     root secret; encoding belongs to the backend)
 *   - any persistence — the phrase is held only in component state
 *     and is never written to localStorage / sessionStorage / cookies
 *   - any logging — callers must not pass the phrase to console.*
 */
import wordlistRaw from "./bip39-wordlist.txt?raw";

export const PHRASE_WORD_COUNT = 24;
export const ROOT_SECRET_BYTES = 32;
const BITS_PER_WORD = 11;
const WORDLIST_LEN = 2048;

export const WORDLIST: readonly string[] = (() => {
  const words = wordlistRaw
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line !== "");
  if (words.length !== WORDLIST_LEN) {
    throw new Error(
      `recoveryPhrase wordlist must be exactly ${WORDLIST_LEN} entries (got ${words.length})`,
    );
  }
  return words;
})();

const WORD_INDEX: ReadonlyMap<string, number> = new Map(
  WORDLIST.map((word, idx) => [word, idx] as const),
);

export type DecodeError =
  | { kind: "wrong_length"; got: number }
  | { kind: "unknown_word"; position: number; word: string }
  | { kind: "bad_checksum" };

export interface DecodeSuccess {
  ok: true;
}

export interface DecodeFailure {
  ok: false;
  error: DecodeError;
}

export type DecodeResult = DecodeSuccess | DecodeFailure;

/**
 * Split a free-form phrase string into trimmed lowercase tokens.
 * Used by the paste handler so an operator can paste 24 words
 * separated by spaces, tabs, or newlines.
 */
export function splitPhraseInput(raw: string): string[] {
  return raw
    .split(/\s+/)
    .map((word) => word.trim().toLowerCase())
    .filter((word) => word !== "");
}

export function isWordInList(word: string): boolean {
  return WORD_INDEX.has(word.trim().toLowerCase());
}

/**
 * Full validation: 24 words, all in the wordlist, checksum byte
 * matches `SHA256(rootSecret)[0]`. The result is a discriminated
 * union rather than an exception so the page can render specific
 * messages ("word N is not recognised", "checksum invalid")
 * without try/catch noise; the variants mirror the backend
 * `RecoveryPhraseError` messages for parity.
 *
 * Async because SubtleCrypto.digest is async. The page awaits
 * this once on submit (not on every keystroke). The derived
 * 32-byte root secret is dropped before this function returns —
 * the frontend never holds it; the backend re-derives it from
 * the phrase string when /auth/recover runs.
 */
export async function verifyPhrase(
  words: readonly string[],
): Promise<DecodeResult> {
  if (words.length !== PHRASE_WORD_COUNT) {
    return { ok: false, error: { kind: "wrong_length", got: words.length } };
  }

  // Re-pack to recover both entropy and the embedded checksum byte.
  let bits = 0n;
  for (let position = 0; position < PHRASE_WORD_COUNT; position++) {
    const word = words[position]!.trim().toLowerCase();
    const idx = WORD_INDEX.get(word);
    if (idx === undefined) {
      return {
        ok: false,
        error: { kind: "unknown_word", position: position + 1, word },
      };
    }
    bits = (bits << BigInt(BITS_PER_WORD)) | BigInt(idx);
  }
  const embeddedChecksum = Number(bits & 0xffn);
  let entropyInt = bits >> 8n;

  const rootSecret = new Uint8Array(ROOT_SECRET_BYTES);
  for (let i = ROOT_SECRET_BYTES - 1; i >= 0; i--) {
    rootSecret[i] = Number(entropyInt & 0xffn);
    entropyInt >>= 8n;
  }

  const digest = new Uint8Array(
    await crypto.subtle.digest("SHA-256", rootSecret),
  );
  // Zero the local copy regardless of outcome — there is no need
  // for the root secret to outlive this function.
  try {
    if (digest[0] !== embeddedChecksum) {
      return { ok: false, error: { kind: "bad_checksum" } };
    }
    return { ok: true };
  } finally {
    rootSecret.fill(0);
  }
}

File: app/security/recovery_phrase.py
"""Recovery phrase encoding for the local secret bootstrap (#114 / ADR-0003).

A recovery phrase is a 24-word encoding of a 32-byte root secret. It uses
the BIP39 English wordlist (vendored under ``wordlist_english.txt``) but
this module deliberately does not expose any "BIP39" / "mnemonic" / "seed
phrase" naming -- the operator-facing concept is "recovery phrase" and
nothing else, per ADR-0003.

Encoding:
  * 256 bits of entropy (the root secret) + 8-bit checksum
    (``SHA256(entropy)[0]``) = 264 bits
  * Split into 24 11-bit groups, each indexing the 2048-word list

The checksum lets us reject typos locally before any HKDF/decryption
attempt -- a tampered phrase fails ``decode_phrase`` with a clear error
rather than silently producing garbage entropy.

Pure functions, no I/O beyond loading the vendored wordlist file once at
import time.
"""

from __future__ import annotations

import hashlib
from importlib.resources import files

ROOT_SECRET_LEN = 32
PHRASE_WORD_COUNT = 24
_BITS_PER_WORD = 11
_WORDLIST_LEN = 2048


class RecoveryPhraseError(ValueError):
    """Raised when a recovery phrase fails validation."""


def _load_wordlist() -> tuple[list[str], dict[str, int]]:
    raw = files("app.security").joinpath("wordlist_english.txt").read_text(encoding="utf-8")
    words = [line.strip() for line in raw.splitlines() if line.strip()]
    if len(words) != _WORDLIST_LEN:
        raise RuntimeError(f"recovery_phrase wordlist must be exactly {_WORDLIST_LEN} entries (got {len(words)})")
    return words, {word: idx for idx, word in enumerate(words)}


_WORDLIST, _WORD_INDEX = _load_wordlist()


def encode_phrase(root_secret: bytes) -> list[str]:
    """Encode a 32-byte root secret as a 24-word recovery phrase."""
    if len(root_secret) != ROOT_SECRET_LEN:
        raise RecoveryPhraseError(f"root secret must be exactly {ROOT_SECRET_LEN} bytes (got {len(root_secret)})")
    checksum = hashlib.sha256(root_secret).digest()[0]
    # 256 bits of entropy + 8 bits of checksum = 264 bits, packed big-endian
    bits = int.from_bytes(root_secret, "big") << 8 | checksum
    words: list[str] = []
    for i in range(PHRASE_WORD_COUNT):
        shift = (PHRASE_WORD_COUNT - 1 - i) * _BITS_PER_WORD
        idx = (bits >> shift) & (_WORDLIST_LEN - 1)
        words.append(_WORDLIST[idx])
    return words


def decode_phrase(phrase: list[str] | str) -> bytes:
    """Decode a recovery phrase back to its 32-byte root secret.

    Accepts either a list of 24 words or a single whitespace-separated
    string. Raises :class:`RecoveryPhraseError` on any validation failure
    (wrong length, unknown word, bad checksum).
    """
    if isinstance(phrase, str):
        words = phrase.strip().split()
    else:
        words = [w.strip().lower() for w in phrase]
    if len(words) != PHRASE_WORD_COUNT:
        raise RecoveryPhraseError(f"recovery phrase must be exactly {PHRASE_WORD_COUNT} words (got {len(words)})")
    bits = 0
    for position, word in enumerate(words, start=1):
        idx = _WORD_INDEX.get(word.lower())
        if idx is None:
            raise RecoveryPhraseError(f"word {position} is not recognised")
        bits = (bits << _BITS_PER_WORD) | idx
    # 264 bits total: top 256 are entropy, bottom 8 are the checksum
    checksum = bits & 0xFF
    entropy_int = bits >> 8
    root_secret = entropy_int.to_bytes(ROOT_SECRET_LEN, "big")
    expected = hashlib.sha256(root_secret).digest()[0]
    if checksum != expected:
        raise RecoveryPhraseError("recovery phrase checksum is invalid")
    return root_secret


def wordlist() -> list[str]:
    """Return a copy of the wordlist (used by tests and the frontend port)."""
    return list(_WORDLIST)
